lit_gpt/utils.py

```diff
@@ -347,3 +359,33 @@ def estimate_flops(model: "GPT", training: bool) -> int:
     # forward + backward
     frozen_ops_per_step = 2 if training else 1
     return ops_per_step * trainable_flops + frozen_ops_per_step * frozen_flops
+
+
+class CycleIterator:
+    """An iterator that cycles through an iterable indefinitely.
+
+    Example:
+        >>> iterator = CycleIterator([1, 2, 3])
+        >>> [next(iterator) for _ in range(5)]
+        [1, 2, 3, 1, 2]
+
+    Note:
+        Unlike ``itertools.cycle``, this iterator does not cache the values of the iterable.
+    """
+    def __init__(self, iterable: Iterable) -> None:
+        self.iterable = iterable
+        self.epoch = 0
+        self._iterator = None
+
+    def __next__(self) -> Any:
+        if self._iterator is None:
+            self._iterator = iter(self.iterable)
```

Comment:
Shouldn't you do this in `__iter__`?

The current way means the iterator cannot be recreated

Reply:
By it's design, I want to be able to directly call `next(CycleIterator())` without calling iter first. The `__iter__` method is just there for completeness.

Reply:
To recreated a second such iterator, you would do `CycleIterator(iterable)` again.

In this sense, `CycleIterator` can be seen as a pure replacement for `iter`.